#  CONTENTS LIST
## 1. Importing libraries
## 2. Import dataframes
## 3. rename columns
## 4. Check for duplicated rows
## 5. Check for missing values
## 6. import ords_prods_merge_aggregation.pkl
## 7. converting user_id into str
## 8. Merge the dataframes on 'user_id'
## 9. Export dataframes

## 3.

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [2]:
#create path
path = r'C:\Users\Christian\Desktop\Mein Ordner\CF\Data Immersion\Achievement4\Instacart Basket Analysis'

In [3]:
# Import customers.csv as df
df = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'customers.csv'))

## 4.

In [4]:
df.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
# rename column names
df.rename(columns={'Surnam': 'surname', 'fam_status': 'family_status', 'STATE': 'State', 'n_dependants': 'dependants',
                   'First Name': 'first_name'}, inplace=True)

In [6]:
df.head()

user_id first_name   surname  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   dependants family_status  income  
0           3       married  165665  
1           0        single   59285  
2           2       married   99568  
3           0        single   42049  
4           1       married   40374

## 5.

In [7]:
# Check for duplicated rows
duplicated = df[df.duplicated()]

In [8]:
duplicated

Empty DataFrame
Columns: [user_id, first_name, surname, Gender, State, Age, date_joined, dependants, family_status, income]
Index: []

In [9]:
# Check for missing values
df.isnull().sum()

user_id              0
first_name       11259
surname              0
Gender               0
State                0
Age                  0
date_joined          0
dependants           0
family_status        0
income               0
dtype: int64

In [10]:
# Filter rows with missing first names
df[df['first_name'].isnull()]


user_id first_name   surname  Gender         State  Age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        dependants family_status  income  
53               2       married   41709  
73               0        single   82518  
82               3       married  117099  
99               1       married  155673  
105              3       married  151819  
...            ...           ...     ...  
206038           3       married   87783  
206044           2       married  108488  
206090           1       married   45275  
206162           0        single   56325  
206171           2       married   59222  

[11259 rows x 10 columns]

All missing values are in the first_name column. Since we can't recrate 11259 first names we leave the cplumn unchanged.

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   user_id        206209 non-null  int64 
 1   first_name     194950 non-null  object
 2   surname        206209 non-null  object
 3   Gender         206209 non-null  object
 4   State          206209 non-null  object
 5   Age            206209 non-null  int64 
 6   date_joined    206209 non-null  object
 7   dependants     206209 non-null  int64 
 8   family_status  206209 non-null  object
 9   income         206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


## 6.

In [12]:
# import ords_prods_merge_aggregation.pkl
df_2 = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data', 'ords_prods_merge_aggregation.pkl'))

In [13]:
df_2.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

     price_range_loc     busiest_day    busiest_days  busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy         Average orders   
1  Mid-range product  Regularly busy    Slowest days         Average orders   
2  Mid-range product  Regularly busy    Slowest days            Most orders   
3  Mid-range product      Least busy    Slowest days         Average orders   
4  Mid-range product      Least busy    Slowest days            Most orders   

  max_order  loyalty_flag average_spend spender_performance frequency_order  \
0        10  New customer      6.367797         Low spender            20.5   
1        10  New customer      6.367797         Low spender            20.5   
2        10  New customer      6.367797         Low spender            20.5   
3        10  New customer      6.367797         Low spender            20.5   
4        10  New customer      6.367797         Low spender            20.5   

          frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer  

[5 rows x 25 columns]

In [14]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   eval_set                object  
 3   order_number            int64   
 4   order_dow               int64   
 5   order_hour_of_day       int64   
 6   days_since_prior_order  float64 
 7   product_id              int64   
 8   add_to_cart_order       int64   
 9   reordered               int64   
 10  product_name            object  
 11  aisle_id                int64   
 12  department_id           int64   
 13  prices                  float64 
 14  _merge                  category
 15  price_range_loc         object  
 16  busiest_day             object  
 17  busiest_days            object  
 18  busiest_period_of_day   object  
 19  max_order               int64   
 20  loyalty_flag            object  
 21  averag

In [15]:
# drop _merge column
df_3 = df_2.drop('_merge', axis=1)

In [16]:
df_3.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

     price_range_loc     busiest_day    busiest_days  busiest_period_of_day  \
0  Mid-range product  Regularly busy  Regularly busy         Average orders   
1  Mid-range product  Regularly busy    Slowest days         Average orders   
2  Mid-range product  Regularly busy    Slowest days            Most orders   
3  Mid-range product      Least busy    Slowest days         Average orders   
4  Mid-range product      Least busy    Slowest days            Most orders   

  max_order  loyalty_flag average_spend spender_performance  frequency_order  \
0        10  New customer      6.367797         Low spender             20.5   
1        10  New customer      6.367797         Low spender             20.5   
2        10  New customer      6.367797         Low spender             20.5   
3        10  New customer      6.367797         Low spender             20.5   
4        10  New customer      6.367797         Low spender             20.5   

          frequency_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer  

[5 rows x 24 columns]

In [ ]:
# converting user_id into str for better memory usage

In [17]:
df_3['user_id'] = df_3['user_id'].astype(str)

In [19]:
df_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 object 
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
 7   product_id              int64  
 8   add_to_cart_order       int64  
 9   reordered               int64  
 10  product_name            object 
 11  aisle_id                int64  
 12  department_id           int64  
 13  prices                  float64
 14  price_range_loc         object 
 15  busiest_day             object 
 16  busiest_days            object 
 17  busiest_period_of_day   object 
 18  max_order               int64  
 19  loyalty_flag            object 
 20  average_spend           float64
 21  spender_performance     objec

In [18]:
df['user_id'] = df['user_id'].astype(str)

In [20]:
# Merge the two dataframes on 'user_id'
df_merge = df_3.merge(df, on='user_id', indicator=True)

In [21]:
df_merge.head()

order_id user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329       1    prior             1          2                  8   
1   2398795       1    prior             2          3                  7   
2    473747       1    prior             3          3                 12   
3   2254736       1    prior             4          4                  7   
4    431534       1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

  first_name  surname  Gender    State Age date_joined dependants  \
0      Linda   Nguyen  Female  Alabama  31   2/17/2019          3   
1      Linda   Nguyen  Female  Alabama  31   2/17/2019          3   
2      Linda   Nguyen  Female  Alabama  31   2/17/2019          3   
3      Linda   Nguyen  Female  Alabama  31   2/17/2019          3   
4      Linda   Nguyen  Female  Alabama  31   2/17/2019          3   

  family_status  income _merge  
0       married   40423   both  
1       married   40423   both  
2       married   40423   both  
3       married   40423   both  
4       married   40423   both  

[5 rows x 34 columns]

In [22]:
# Export the merged DataFrame as a pickle file
df_merge.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_customer.pkl'))
